Add checkheader tool

qmuntal opened the pull request:
Checkheader is a standalone tool which, given the header file where we define all OpenSSL functions, generates a new C header file that statically checks that our definitions and the OpenSSL definitions existing in the machine match 1 to 1.

This will avoid issues like [#17](https://github.com/microsoft/go-crypto-openssl/pull/17) or #5.

The basic idea is that it translates `DEFINEFUNC(int, RAND_bytes, (uint8_t *a0, size_t a1), (a0, a1))` into `__typeof__(int(*)(uint8_t *, size_t)) __check_0 = RAND_bytes;`.
It is also smart enough to honor  the version requirements of version-aware `DEFINEFUNC` macros.

The typical usage would be something like:
```bash
$ sh ./scripts/openssl.sh 1.0.2 // Install OpenSSL binaries and development headers
$ go run ./cmd/checkheader --ossl_include /usr/local/src/openssl-1.0.2/include ./openssl/shims.h
```

dagood reviewed (commented):
Did a pass through the docs and comments to make sure this tool will make sense to everyone, now that we're aiming to add it to a repo with more eyes on it. 🙂

qmuntal commented on cmd/checkheader/main.go:
```diff
@@ -0,0 +1,302 @@
+
+const description = `
+Example: A check operation:
+  go run ./cmd/checkheader --ossl-include /usr/local/src/openssl-1.1.1/include ./openssl/shims.h 
```
Indeed, I'll pass it using `-shim` so it's better documented.

derekparker reviewed (approved):
LGTM

qmuntal commented:
I've rebased this branch and added a new CI step that verifies that FIPS is really enabled when we call `SetFIPS(true)`: https://github.com/golang-fips/openssl-fips/pull/18/commits/e9275ad7c17af39e57941a6e38883462e44dcdc4

ueno commented:
Sorry for the belated comment, but this looks great! Thank you for working on this @qmuntal :-)
